{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "demanding-space",
   "metadata": {},
   "source": [
    "# Introduction to CPTAC\n",
    "This tutorial will introduce you to the CPTAC dataset through a Python package called `cptac` that was written to access the CPTAC data. This package is similar to `TCGAbiolinks`. The information in `cptac` is stored using `pandas` dataframes. For those curious, the CPTAC documentation is contained [here](https://pypi.org/project/cptac/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "thick-choice",
   "metadata": {},
   "source": [
    "### Installing and Importing `cptac`\n",
    "Just as in R, we need to install before we can load in `cptac` to our current python environment. If you have not done so already, we need to have installed the package (you can check if it's installed by running the `import cptac` code block below). If you get an error, you will have to take the following steps:\n",
    "\n",
    "1. Shut down your jupyter notebook kernel in the terminal with CTRL+C.\n",
    "2. Make sure that you have deactivated your base environment and activated your `qbio_student` environment.\n",
    "3. Run `pip install cptac`.\n",
    "4. Open the tutorial again through jupyter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "viral-shelf",
   "metadata": {},
   "source": [
    "### Start exploring CPTAC with `cptac`\n",
    "Similar to `TCGAbiolinks`, we need to load the package and download the data:\n",
    "\n",
    "1. Import the package (`cptac`).\n",
    "2. Look at the data sets available to us with the `list_datasets()` function. As a reminder, remember the package_name.function() syntax!\n",
    "3. Download the one of interest (named `Colon`).\n",
    "4. Load it into our python environment.\n",
    "\n",
    "You might get an error message warning that a method is deprecated -- you can ignore this!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d34ed3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 1. Import cptac\n",
    "\n",
    "\n",
    "# 2. Examine the data sets available with list_datasets()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exceptional-integration",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Download the colorectal data set -- fill in dataset!\n",
    "cptac.download(dataset=...)\n",
    "\n",
    "# 4. Load the colon cancer data set\n",
    "colon = cptac.Colon()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-growth",
   "metadata": {},
   "source": [
    "Now, we have our `colon` object containing a bunch of data. We can use the `list_data()` function the data available for the breast cancer dataset. \n",
    "\n",
    "Notice that in addition to the proteomics data, we can also accesss the accompanying transcriptomics, CNV, etc. Additionally, we will focus on the broad proteomics data; however, acetylproteomics and phosphoproteomics are also interesting aspects to explore. These data examine \"post-translational modifications\", in this case the addition of acetyl or phosphate groups to proteins after they are translated, respectively. This data can give insight into which proteins are being used in which cellular pathways.\n",
    "\n",
    "Finally, one last caveat with the data is that many proteins can't be identified very well, so they won't appear as a column (for example, APC isn't in the proteomics data set)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "857a760c",
   "metadata": {},
   "source": [
    "**Exercise**. Call the `list_data()` function from the `colon` object to examine the different data contained in the `colon` object. Much like calling functions from a package, we can use the `colon.function()` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71758797",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the list_data() function here\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "323f288d",
   "metadata": {},
   "source": [
    "**Exercise.** Now, we want to access the proteomics data. Use the `get_proteomics()` function to store the proteomics data into a separate data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abf63378",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_data = ... # get the proteomics data\n",
    "protein_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "changed-leather",
   "metadata": {},
   "source": [
    "## Analysis\n",
    "We'll be analyzing the proteomics data with `numpy` and `pandas`, so we'll have to load them in as usual.\n",
    "\n",
    "**Exercise**. Import `numpy` and `pandas`. Be sure to assign them their usual shortened names!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adapted-union",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "promotional-ivory",
   "metadata": {},
   "source": [
    "### Short Pandas Review\n",
    "We can access the rows/column names of a data frame in two ways:\n",
    "\n",
    "1. We can get the row and column names together in a list with the `axes` attribute.\n",
    "2. We can get the row or column names individually in the `index` (row) or `columns` (column) attribute.\n",
    "\n",
    "\n",
    "**Exercise.** Print the `axes`, `index`, and `columns` of the `protein_data` object. Then, determine the dimensions of the `protein_data`, stored in the `shape` attribute. Finally, answer the following questions:\n",
    "\n",
    "- How many rows and columns are there in the data frame?\n",
    "- What do the rows and columns represent?\n",
    "- How many patients are there?\n",
    "- What about proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "expired-reach",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Explore the data frame here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add2d694",
   "metadata": {},
   "source": [
    "As for indexing, recall that `loc[]` accesses values by name, and `iloc[]` accesses values by index. `iloc[]` won't be too necessary for this data.\n",
    "\n",
    "**Exercise.** Access the following:\n",
    "1. Print the relative fold change of patient 01CO014 in the `CTNNB1` (beta-catenin) protein. Then, interpret this value -- does this patient have higher or lower protein levels than the reference? How much different is it? (Hint: use `loc[]`. Be sure to specify both the rows and the columns!)\n",
    "2. Access all values in the `COL4A1` column. Because this is a column, there's many ways to do so -- can you think of at least two?\n",
    "3. Finally, use boolean indexing to access the patient IDs with high (â‰¥1) `KRAS` expression:\n",
    "    a. Think about if KRAS expression information will be contained in a row or a column!\n",
    "    b. Create a boolean mask with the row/column that contains the expression information of KRAS.\n",
    "    c. Apply the mask to the row/column names that contains the patient IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda8c90d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Access cell of patient 01CO014, for the gene CTNNB1\n",
    "part1 = ...\n",
    "assert(part1 == -0.0431)\n",
    "\n",
    "# 2. Access all the values in the COL4A1 column\n",
    "part2 = ...\n",
    "assert(len(part2) == 197)\n",
    "\n",
    "# 3. Get the patients with high KRAS expression\n",
    "KRAS_mask = ...\n",
    "part3 = ...\n",
    "assert(part3 == \"05CO028.N\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forbidden-advertiser",
   "metadata": {},
   "source": [
    "### Additional data in cptac\n",
    "You can also access other data using the `cptac` python package for the same patients, using the corresponding `get` function.\n",
    "\n",
    "**Exercise**. Access the RNA (`transcriptomics`) and `clinical` data from the colon object. The function name is very similar in syntax to the `get_proteomics()` function from before!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13e51b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_data = ...\n",
    "clinical_data = ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3e3bd2",
   "metadata": {},
   "source": [
    "**Exercise**. Examine these dataframes. In particular:\n",
    "\n",
    "- What do the values in the `rna_data` data frame represent?\n",
    "- Compare the dimensions of the `protein_data`, `rna_data`, and `clinical_data`. Are they equal?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a54f52a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# explore the data frames!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cac2236",
   "metadata": {},
   "source": [
    "We can use the `intersect1d()` function from `numpy` to see which patient barcodes are shared between the data frames. **The reason we might be interested in these patients is that we can compare the transcriptomics to proteomics data.**\n",
    "\n",
    "1. Use `intersect1d()` to create a list that contains the patient IDs for each pair of data frames as mentioned below.\n",
    "2. Use a for loop to print the lengths of these three data frames, and compare them to the number of rows. The for loop skeleton is below. What do you notice about the three lengths? Do the numbers correspond to anything in particular?\n",
    "3. Finally, identify which intersection contains the patient IDs where there is data for all three levels of data (proteomics, transcriptomics, and clinical). How can we tell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "chemical-location",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.\n",
    "name_intersects = [\n",
    "    , # 0. fill in intersecting for protein/rna here\n",
    "    , # 1. fill in intersecting for protein/clinical here\n",
    "    , # 2. fill in the intersecting for rna/clinical here\n",
    "]\n",
    "\n",
    "# 2. Print the lengths here\n",
    "# write a for loop here\n",
    "    print(f\"The length of ____ is _____\")  # fill in here\n",
    "    \n",
    "# 3. Which comparison contains the patient names that have all three levels?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46124874",
   "metadata": {},
   "source": [
    "### Examining the clinical data\n",
    "Let's explore the clinical data in more detail.\n",
    "\n",
    "**Exercise**. First, let's use `head()` to glance as to what data is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b38bbe6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# call head on the clinical data!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e5486b7",
   "metadata": {},
   "source": [
    "**Exercise**. You might notice that the Age column values don't look like years -- they're in months. Divide the values by 12 in the `clinical_data` data frame to turn them into years. Then, use `head()` again to make sure the ages were converted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61978ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert ages here!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d704ebf5",
   "metadata": {},
   "source": [
    "There are actually control (non-cancer) tissue samples in the data! Let's show this using the `unique()` function from numpy.\n",
    "\n",
    "**Exercise.** Call the `unique()` function on the `Sample_Tumor_Normal` column. The function will return the unique values in the column in a sorted order -- this is super useful for examining categorical variables, like tumor stage and tumor status, for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "416c93c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the levels of the Sample_Tumor_Normal column with unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5ff472",
   "metadata": {},
   "source": [
    "**Exercise.** Let's examine how tumor stage varies as a function of age:\n",
    "\n",
    "1. Import the plotting libraries `matplotlib.pyplot` and `seaborn` (use the standard abbreviations presented last time).\n",
    "2. Create an `age_category` column in `clinical_data`. As before, we'll define \"Young\" as under 50, and \"Old\" as 50 and older. (Hint: use the `where()` function from numpy, the equivalent to `ifelse()`).\n",
    "3.  The \"Stage\" information for these patients is NaN (the pandas version of NA). Use the `isna()` function from pandas and boolean indexing (i.e. `where()` from numpy again) to replace these NaN values in the `Stage` column with \"No cancer\".\n",
    "3. Using nested for loop and boolean indexing, count the number of old and young patients that have No cancer, Stage I, Stage II, Stage III, and Stage IV cancer. \n",
    "4. Compare your results to the barplot. Do your numbers make sense?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e84c387",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Import libraries here\n",
    "\n",
    "\n",
    "# 2. Create the age_category column in clinical_data\n",
    "\n",
    "\n",
    "# 3. Replace NaN with No cancer\n",
    "\n",
    "\n",
    "\n",
    "## this will programatically get all the stages in order\n",
    "stage_categories = np.unique(stage_categories)\n",
    "assert(np.all(stage_categories == ['No cancer', 'Stage I', 'Stage II', 'Stage III', 'Stage IV']))\n",
    "\n",
    "# 4. Loop through all the stage categories and count\n",
    "\n",
    "for age_cat in [\"Old\", \"Young\"]: # first get the old data, then the young data\n",
    "    print(age_cat)\n",
    "    # fill this in with the cancer category:\n",
    "    for ...:\n",
    "        # 1. create a subset of the data frame with the old/young patients\n",
    "        # 2. count the number of patients with the stage of cancer and print\n",
    "\n",
    "\n",
    "        \n",
    "# 5. A sample barplot to compare your results\n",
    "clinical_data_copy = clinical_data\n",
    "clinical_data_copy[\"Stage\"] = np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "sns.countplot(\n",
    "    x = \"age_category\",\n",
    "    hue = \"Stage\",\n",
    "    hue_order = [\"No cancer\", \"Stage I\", \"Stage II\", \"Stage III\", \"Stage IV\"],\n",
    "    data = clinical_data_copy\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4747fb4c",
   "metadata": {},
   "source": [
    "## Comparing Transcriptomic and Proteomic Data\n",
    "A natural question to ask is if the levels of RNA expression correlate with protein expression. We can explore this question with the help of the `scipy` library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5e207ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats # we are using the stats package in particular"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1395e3d7",
   "metadata": {},
   "source": [
    "The first thing we need to do is to identify which patients and genes are shared between the transcriptomic and proteomic data sets. We need to identify which genes are shared, but luckily, we already got that data before for the patients! We will use this data to create two new data frames that contain only the rows and column names that are shared between these two data sets. \n",
    "\n",
    "**Exercise**. To make sure the data frames match:\n",
    "1. Identify the names of the genes that are shared between the two datasets (refer to the exercise when we created the dictionary). Is this data the row or column names?\n",
    "2. Create the `rna_shared` and `prot_shared` data frames as described above (Hint: how can we access rows/columns by name?).\n",
    "3. Finally, we need to make sure all the rows and columns are in the correct order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2589e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Identify the genes (RNA, protein) shared between the two data sets \n",
    "shared_rna_prot = ...\n",
    "\n",
    "# 2. Create the two data frames\n",
    "rna_shared = ...\n",
    "prot_shared = ...\n",
    "\n",
    "# 3. make sure that everything is in the correct order\n",
    "assert(np.all(rna_shared.index == prot_shared.index))\n",
    "assert(np.all(rna_shared.columns == prot_shared.columns))\n",
    "\n",
    "# you can, in fact, chain comparisons in python\n",
    "# DO NOT DO THIS IN ANY OTHER LANGUAGE\n",
    "assert(rna_shared.shape == prot_shared.shape == (96, 7572))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "769b1472",
   "metadata": {},
   "source": [
    "Now, we can see how correlated the RNA and protein levels are. We'll use Spearman correlation from the `stats` library, which is `spearmanr()`. For example, to get the correlation in the KRAS gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d7aff46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we need the nan_policy=\"omit\" to throw out NaN values\n",
    "corr, pval = stats.spearmanr(rna_shared[\"KRAS\"], prot_shared[\"KRAS\"], nan_policy=\"omit\")\n",
    "\n",
    "# round() to truncate the numbers\n",
    "print(f\"The correlation is {round(corr, 3)} (p = {round(pval, 6)}).\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3174183c",
   "metadata": {},
   "source": [
    "**Exercise**. Investigate the correlation between the RNA and the protein data for the CD8A gene, printing the values. Then, using seaborn, create a scatterplot, with RNA on the x-axis and protein on the y-axis (be sure to create your fig/ax!). Does the correlation look like it makes sense? As usual, be sure to label the axes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1002dc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"CD8A\" # this will save you some typing and make the code more flexible for exploring other genes\n",
    "\n",
    "# get correlation and print\n",
    "\n",
    "# create scatter plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "556dffec",
   "metadata": {},
   "source": [
    "## Heatmaps\n",
    "A problem arises if we want to compare many correlations. Heatmaps are useful for visualizing a large number of comparisons. To make a heatmap, we'll use the `heatmap()` function from seaborn (the `cluster()` function is closely related; the difference is that `clustermap()` will reorganize the rows/columns so that similar rows/columns are grouped near each other). Let's do the following: \n",
    "\n",
    "1. Set up our data frame to hold all comparisons. All you need to do is to access the first 20 gene names.\n",
    "2. Calculate the correlations for the first 20 genes (just to save time). You'll need to use two for loops.\n",
    "3. Call `clustermap()` -- this is filled in for you!\n",
    "4. Interpret the data. There is a \"light\" diagonal along the heatmap from the top-left to bottom-right. Is this expected? Why or why not?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a95838ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncomparisons = 20 # define this variable in case we want to change the number of correlations to test\n",
    "                  # this makes it less likely you'll forget to change a number, e.g. in the data frame shape\n",
    "gene_names = ... # get the first ncomparisons gene names\n",
    "\n",
    "\n",
    "# Don't worry about this code\n",
    "# It's good practice to declare your data frame beforehand (it's much faster than appending to a list)\n",
    "# We fill everything in with 0 just as a placeholder\n",
    "corr_df = pd.DataFrame(np.ndarray(shape=(ncomparisons, ncomparisons), dtype=np.float16),\n",
    "                      index = gene_names,\n",
    "                      columns = gene_names)\n",
    "\n",
    "# 2. fill in the data frame!\n",
    "for g1 in gene_names:\n",
    "    for g2 in gene_names:\n",
    "        # calculate the correlations between protein and RNA\n",
    "        # then, use .loc[] to store the correlation in corr_df\n",
    "\n",
    "# 3. create the heat map\n",
    "sns.heatmap(\n",
    "    corr_df,\n",
    "    cmap='mako'\n",
    ")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# 4. interpret!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "attached-charity",
   "metadata": {},
   "source": [
    "### Saving pandas dataframes\n",
    "While you can certainly reload in the data every time you wish to use it, we also want to make sure you know how to save and read in a pandas dataframes. We will utilize the `to_csv()` function and `read_csv()` function. Notice in the `read_csv()` function we utilize the `index_col` argument. To observe what this argument does, try removing it and looking at the dataframe. Notice what the index values are. \n",
    "\n",
    "**Exercise**. To save our data:\n",
    "1. Make a folder in your `analysis_data` folder called `cptac`.\n",
    "2. Edit the path to save into `analysis_data`\n",
    "3. Also save the RNA and clinical data as well to the same directory (with the same name convention)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "resistant-approach",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# RENAME the path to analysis_data!\n",
    "protein_data.to_csv(\"YOUR_PATH/analysis_data/cptac/cptac_protein.csv\")\n",
    "protein_data_read_example = pd.read_csv(\"PATH/Filename.csv\", index_col=0)\n",
    "protein_data_read_example\n",
    "\n",
    "# Repeat with clinical and RNA"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
